File: openharmony/arkui_napi/utils/log.cpp
/*
 * Copyright (c) 2021-2022 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "utils/log.h"

#include <cstdint>

[[maybe_unused]] static void StripFormatString(const std::string& prefix, std::string& str)
{
    for (auto pos = str.find(prefix, 0); pos != std::string::npos; pos = str.find(prefix, pos)) {
        str.erase(pos, prefix.size());
    }
}

#if defined(ANDROID_PLATFORM)

#include <android/log.h>

#define LOG_TAG "NAPI"

constexpr int LOG_LEVEL[] = { ANDROID_LOG_DEBUG, ANDROID_LOG_INFO, ANDROID_LOG_WARN, ANDROID_LOG_ERROR,
    ANDROID_LOG_FATAL };

NAPI_EXPORT void PrintLog(LogLevel level, const char* fmt, ...)
{
    std::string newFmt(fmt);
    StripFormatString("{public}", newFmt);
    StripFormatString("{private}", newFmt);
    va_list args;
    va_start(args, fmt);
    __android_log_vprint(LOG_LEVEL[static_cast<int>(level)], LOG_TAG, newFmt.c_str(), args);
    va_end(args);
}

#elif defined(MAC_PLATFORM) || defined(WINDOWS_PLATFORM) || defined(IOS_PLATFORM) || defined(LINUX_PLATFORM)

#include <securec.h>

constexpr uint32_t MAX_BUFFER_SIZE = 4096;

NAPI_EXPORT void PrintLog(LogLevel level, const char* fmt, ...)
{
    std::string newFmt(fmt);
    StripFormatString("{public}", newFmt);
    StripFormatString("{private}", newFmt);

    va_list args;
    va_start(args, fmt);

    char buf[MAX_BUFFER_SIZE] = { '\0' };
    int ret = vsnprintf_s(buf, sizeof(buf), sizeof(buf) - 1, newFmt.c_str(), args);
    va_end(args);
    if (ret < 0) {
        return;
    }

    printf("%s\r\n", buf);
    fflush(stdout);
}
#endif

File: openharmony/arkui_ace_engine/frameworks/bridge/arkts_frontend/koala_projects/arkoala/ohos-sdk/scripts/ohos-app/log.mjs
/*
 * Copyright (c) 2022-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ohProjectConf } from "./../ohconf.mjs"
import { execCmd, lazy, platformPath } from "../utils.mjs"

const LZ_OHCONF = lazy(ohConfPath => ohProjectConf(ohConfPath))
const LZ_HDC = lazy(() => LZ_OHCONF().hdcPath())
const LZ_HDC_OPTIONS = lazy(hdcExtraOpts => hdcExtraOpts ? hdcExtraOpts.split(" ") : [])

async function execHDC(args, options = {}) {
    return execCmd(LZ_HDC(), [...LZ_HDC_OPTIONS(), ...args], {
        ...options,
        instantLogging: true,
    })
}

export async function getPID(bundleName = LZ_OHCONF().bundleName()) {
    const out = await execHDC(["shell", `ps -ef | awk '$NF=="${bundleName}" {print $2}'`])
    return out  
}

export async function logByPID(pid) {
    await execHDC(["shell", `hilog -P ${pid}`])
}

export async function allLog() {
    await execHDC(["shell", "hilog"])
}

export async function log(options) {
    const { all, bundleName } = options
    if (all) {
        await allLog()
    } else {
        const pid = await getPID(bundleName)
        await logByPID(pid)
    }
}